### ISSUE TITLE
[WIN] Parallel tests execution fails because of locked files

### ISSUE THREAD

**Issue by gshimansky** [intel/intel-xpu-backend-for-triton#2777]
### Describe the bug

Trying to execute multiple unit tests in parallel with xdist `-n X` on Windows leads to failures. It happens most likely because one worker compiles and starts executing a kernel using a launcher DLL (with `pyd` extension on windows) from ~/.triton/cache folder while another worker tries to compile the same kernel and write the a launcher DLL into the same folder. On Windows a DLL that is loaded into a process is locked and cannot be modified, so 2nd worker that tries to write a .pyd file gets an IO error and fails. This doesn't happen on Linux because Linux doesn't lock files that are open by running processes.

Any ideas on how to reliably solve this are welcome.

### Environment details

Triton on any GPU running on Windows.

**Comment by anmyachev** [comment]
Maybe try to implement pytest' session-level fixture with lock files, in which the launcher will be built only once?
For ref: https://pytest-xdist.readthedocs.io/en/stable/how-to.html#making-session-scoped-fixtures-execute-only-once

**Comment by gshimansky** [comment]
Ok it looks like the problem is with how os.replace is implemented on Windows. It is not atomic and consists of two system calls: SetRenameInformationFile and CloseFile between which if some other process tries to open this file (for reading or writing doesn't matter) it gets a SHARING VIOLATION error. I wrote two small tests that demonstrate this behavior:

```python
import os

count = 0
while True:
    name = f"{count}.txt"
    f = open(name, "w")
    f.write("test\n")
    f.close()

    try:
        os.replace(name, "test.txt")
    except PermissionError:
        print(f"Failed with {name}")

    count += 1
```
```python
from pathlib import Path

count = 0
while True:
    try:
        line = Path("test.txt").read_text()
    except PermissionError:
        print(f"Failed {count}")
    count += 1
```
When they run together the second program produces errors that we see in triton cache.

**Comment by gshimansky** [comment]
Also what should be taken into account is that VS Code IDE often opens and locks files because it monitors filesystem for changes. So if you are running the above tests make sure that you are not running VS Code (Code.exe process name).

**Comment by gshimansky** [comment]
Trace from Process Monitor that demonstrates the contention. Process 95332 renames temp file in two system calls and process 199284 tries to read it between these calls, gets an error.
![image](https://github.com/user-attachments/assets/ed91e1b0-60f7-4111-8a33-9bf1169b8e77)

**Comment by alexbaden** [comment]
If you remove the DLL from the equation (maybe by moving the directory?) do you have this issue with the cached IR and generated device code?

**Comment by pbchekin** [comment]
Interestingly, there are no such errors in this run: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/12355194841/job/34478248043. The only PermissionError i noticed is #3019, which is simple to fix.

**Comment by gshimansky** [comment]
> Interestingly, there are no such errors in this run: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/12355194841/job/34478248043. The only PermissionError i noticed is #3019, which is simple to fix.

These exceptions are pretty rare, they happen about 2-4 times per 10k tests in core test suite when I run them on 16 workers. You used only 2 workers and probably just were lucky this time.

**Comment by pbchekin** [comment]
> > Interestingly, there are no such errors in this run: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/12355194841/job/34478248043. The only PermissionError i noticed is #3019, which is simple to fix.
> 
> These exceptions are pretty rare, they happen about 2-4 times per 10k tests in core test suite when I run them on 16 workers. You used only 2 workers and probably just were lucky this time.

Right, I am re-running with 16 workers, trying to reproduce.

**Comment by pbchekin** [comment]
Reproduced in https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/12361948103.
```
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\vagrant\\.triton\\cache\\64ML3ZM6GIJOPBOHKJ7BC3NE5FSANALHDMOW36QYF4ILQMPA2BUA\\tmp.pid_26440_47b27b34-b36a-4e1c-911c-7666b42124c5\\kernel_broadcast_rhs.ttir' -> 'C:\\Users\\vagrant\\.triton\\cache\\64ML3ZM6GIJOPBOHKJ7BC3NE5FSANALHDMOW36QYF4ILQMPA2BUA\\kernel_broadcast_rhs.ttir'
```

**Comment by gshimansky** [comment]
> Reproduced in https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/12361948103.
> 
> ```
> PermissionError: [WinError 5] Access is denied: 'C:\\Users\\vagrant\\.triton\\cache\\64ML3ZM6GIJOPBOHKJ7BC3NE5FSANALHDMOW36QYF4ILQMPA2BUA\\tmp.pid_26440_47b27b34-b36a-4e1c-911c-7666b42124c5\\kernel_broadcast_rhs.ttir' -> 'C:\\Users\\vagrant\\.triton\\cache\\64ML3ZM6GIJOPBOHKJ7BC3NE5FSANALHDMOW36QYF4ILQMPA2BUA\\kernel_broadcast_rhs.ttir'
> ```

Here you are getting an error when two processes contend on `os.replace` function call. This is a much more common error and happens for me in hundreds of cases when run on multiple workers. I've fixed this error in this patch https://github.com/intel/intel-xpu-backend-for-triton/pull/2787/files. This patch however doesn't fix a problem when one worker executes `os.replace` and another tries to open the same file for reading.

**Comment by pbchekin** [comment]
No PermissionErrors in the latest run: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/12383767088/job/34567109930

Potentially the following combination of changes helped:
* Triton cache dir per pytest worker: https://github.com/intel/intel-xpu-backend-for-triton/commit/58c6ccc6db03c9b2ce0ad48e08d1a710e5beb8fd (will go in a separate PR)
* #2787 
* #3020 
* #3028
